File: examples/flutter_counter/android/app/src/main/kotlin/com/example/flutter_counter/CounterService.kt
package com.example.flutter_counter

import com.itoken.team.nativerpc.core.NativeRPCService
import com.itoken.team.nativerpc.dsl.serviceDefinition

class CounterService : NativeRPCService() {
    private var count = 0
    
    override fun definition() = serviceDefinition {
        Name("counter")
        
        Function0<Int>("getValue") {
            count
        }
        
        Function0<Int>("increment") {
            count++
            emit("countChanged", mapOf("count" to count))
            count
        }
        
        Function0<Int>("decrement") {
            count--
            emit("countChanged", mapOf("count" to count))
            count
        }
        
        Function0<Int>("reset") {
            count = 0
            emit("countChanged", mapOf("count" to count))
            count
        }
        
        Events("countChanged")
    }
}

File: examples/flutter_counter/ios/Runner/CounterService.swift
import Foundation
import NativeRPCKit

class CounterService: NativeRPCService {
    private var count = 0
    
    @ServiceDefinitionBuilder
    override func definition() -> ServiceDefinitionContainer {
        Name("counter")
        
        Function("getValue") { [weak self] () -> Int in
            return self?.count ?? 0
        }
        
        Function("increment") { [weak self] () -> Int in
            guard let self = self else { return 0 }
            self.count += 1
            self.emit("countChanged", data: ["count": self.count])
            return self.count
        }
        
        Function("decrement") { [weak self] () -> Int in
            guard let self = self else { return 0 }
            self.count -= 1
            self.emit("countChanged", data: ["count": self.count])
            return self.count
        }
        
        Function("reset") { [weak self] () -> Int in
            guard let self = self else { return 0 }
            self.count = 0
            self.emit("countChanged", data: ["count": self.count])
            return self.count
        }
        
        Events("countChanged")
    }
}
